core/services/pg/advisory_lock.go

```diff
@@ -27,38 +27,38 @@ type advisoryLock struct {
 	conn          *sqlx.Conn
 	checkInterval time.Duration
 	logger        logger.Logger
-
-	chStop chan struct{}
-	wg     sync.WaitGroup
+	stop          func()
+	wgReleased    sync.WaitGroup
 }
 
 // NewAdvisoryLock returns an advisoryLocker
 func NewAdvisoryLock(db *sqlx.DB, id int64, lggr logger.Logger, checkInterval time.Duration) AdvisoryLock {
-	return &advisoryLock{id, db, nil, checkInterval, lggr.Named("AdvisoryLock").With("advisoryLockID", id), make(chan struct{}), sync.WaitGroup{}}
+	return &advisoryLock{id, db, nil, checkInterval, lggr.Named("AdvisoryLock").With("advisoryLockID", id), func() {}, sync.WaitGroup{}}
 }
 
-// TakeAndHold will block and wait indefinitely until it can get its first lock
+// TakeAndHold will block and wait indefinitely until it can get its first lock or ctx is cancelled.
+// Use Release() function to release the acquired lock.
 // NOT THREAD SAFE
-func (l *advisoryLock) TakeAndHold() (err error) {
+func (l *advisoryLock) TakeAndHold(ctx context.Context) (err error) {
 	l.logger.Debug("Taking initial advisory lock...")
 	retryCount := 0
 
-	ctxStop, cancel := utils.ContextFromChan(l.chStop)
-	defer cancel()
+	lctx, cancel := context.WithCancel(context.Background())
+	l.stop = cancel
```

Comment:
I think it would clear this up if these two lines were moved right down, to be just above `go l.loop(lctx)`.

And, for the queries within the loop, we use a context that inherits from the passed `ctx`.

So, the passed in context applies only to the synchronous operation of TakeAndHold - once the loop is going, you _must_ Release to cancel it.

Reply:
Ah, what a terrible mistake of mine. Fixed.